chartfx-math/src/main/java/de/gsi/math/filter/fir/FirFilter.java

```diff
@@ -0,0 +1,185 @@
+            final double es = Math.sqrt(Math.pow(100 / (100 - rippleP), 2) - 1);
+            final double vx = 1 / numberOfPoles * Math.log(1 / es + Math.sqrt(1 / Math.pow(es, 2) + 1));
+            double kx = 1 / numberOfPoles * Math.log(1 / es + Math.sqrt(1 / Math.pow(es, 2) - 1));
+            kx = (Math.exp(kx) + Math.exp(-kx)) / 2;
+            rp = rp * ((Math.exp(vx) - Math.exp(-vx)) / 2) / kx;
+            ip = ip * ((Math.exp(vx) + Math.exp(-vx)) / 2) / kx;
+        }
+
+        // s-domain to z-domain conversion
+        final double t = 2 * Math.tan(0.5);
+        final double w = 2 * Math.PI * cutoffFraction;
+        final double m = Math.pow(rp, 2) + Math.pow(ip, 2);
+        double d = 4 - 4 * rp * t + m * Math.pow(t, 2);
+        final double x0 = Math.pow(t, 2) / d;
+        final double x1 = 2 * Math.pow(t, 2) / d;
+        final double x2 = Math.pow(t, 2) / d;
+        final double y1 = (8 - 2 * m * Math.pow(t, 2)) / d;
+        final double y2 = (-4 - 4 * rp * t - m * Math.pow(t, 2)) / d;
+        double k = 0;
+
+        // LP to LP, or LP to HP transform
+        switch (filterType) {
+        case HIGH_PASS:
+            k = -Math.cos(0.5 * w + 0.5) / Math.cos(0.5 * w - 0.5);
+            break;
+        case LOW_PASS:
+        default:
+            k = Math.sin(0.5 - 0.5 * w) / Math.sin(0.5 * w + 0.5);
+            break;
+        }
+        d = 1 + y1 * k - y2 * Math.pow(k, 2);
+
+        final double[] filterParameters = new double[5];
+        final double a0 = (x0 - x1 * k + x2 * Math.pow(k, 2)) / d; // a0
+        double a1 = (-2 * x0 * k + x1 + x1 * Math.pow(k, 2) - 2 * x2 * k) / d; // a1
+        final double a2 = (x0 * Math.pow(k, 2) - x1 * k + x2) / d; // a2
+        double b1 = (2 * k + y1 + y1 * Math.pow(k, 2) - 2 * y2 * k) / d; // b1
+        final double b2 = (-Math.pow(k, 2) - y1 * k + y2) / d; // b2
+        if (filterType == FilterType.HIGH_PASS) {
+            a1 = -a1;
+            b1 = -b1;
+        }
+        filterParameters[0] = a0;
+        filterParameters[1] = a1;
+        filterParameters[2] = a2;
+        filterParameters[3] = b1;
+        filterParameters[4] = b2;
+
+        return filterParameters;
+    }
+
+    /**
+     * filter from: "The Scientist and Engineer's Guide to DSP" Chapter 20
+     *
+     * @param signal input signal
+     * @param output output signal (if null a new vector is being allocated)
+     * @param cutoffFraction cutoffFreq must be smaller than half the sample rate
+     * @param filterOrder can be between 2 and 20
+     * @param filterType 0: lowPass 1: highPass
+     * @param ripplePercent ripplePercent is amount of ripple in Chebyshev filter (0-29) (ripplePercent == 0 -&gt;
+     *            Butterworth)
+     * @return filtered output signal
+     */
+    public static double[] filterSignal(final double[] signal, final double[] output, final double cutoffFraction, final int filterOrder, final FilterType filterType, final double ripplePercent) {
```

Comment:
Codacy found an issue: [The method 'filterSignal(double, double, double, int, FilterType, double)' has an NPath complexity of 3200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5764587)